typing/typemod.ml

```diff
@@ -2353,40 +2353,92 @@ let type_package env m p nl =
   let modl = type_module env m in
   let scope = Ctype.create_scope () in
   Typetexp.widen context;
-  let (mp, env) =
-    match modl.mod_desc with
-    | Tmod_ident (mp,_) -> (mp, env)
-    | Tmod_constraint ({mod_desc=Tmod_ident (mp,_)}, _, Tmodtype_implicit, _)
-        -> (mp, env)  (* PR#6982 *)
-    | _ ->
-      let (id, new_env) =
-        Env.enter_module ~scope "%M" Mp_present modl.mod_type env
+  let nl', tl', env =
+    match nl with
+    | [] -> [], [], env
+    | nl ->
+      let type_path, env =
+        match modl.mod_desc with
+        | Tmod_ident (mp,_)
+        | Tmod_constraint
+            ({mod_desc=Tmod_ident (mp,_)}, _, Tmodtype_implicit, _) ->
+          (* We special case these because interactions between
+             strengthening of module types and packages can cause
+             spurious escape errors. See examples from PR#6982 in the
+             testsuite. This can be removed when such issues are
+             fixed. *)
+          let rec path = function
+            | Lident name -> Pdot(mp, name)
+            | Ldot(m, name) -> Pdot(path m, name)
+            | Lapply _ -> assert false
+          in
+          path, env
+        | _ ->
+          let sg = extract_sig_open env modl.mod_loc modl.mod_type in
+          let sg, env = Env.enter_signature ~scope sg env in
+          let types, modules =
+            List.fold_left
+              (fun acc item ->
+                 match item with
+                 | Sig_type(id, _, _, _) ->
+                     let types, modules = acc in
+                     let types = String.Map.add (Ident.name id) id types in
+                     types, modules
+                 | Sig_module(id, _, _, _, _) ->
+                     let types, modules = acc in
+                     let modules = String.Map.add (Ident.name id) id modules in
+                     types, modules
+                 | _ -> acc)
+              (String.Map.empty, String.Map.empty) sg
+          in
+          let rec module_path = function
+            | Lident name -> Pident (String.Map.find name modules)
+            | Ldot(m, name) -> Pdot(module_path m, name)
+            | Lapply _ -> assert false
+          in
+          let type_path = function
+            | Lident name -> Pident (String.Map.find name types)
+            | Ldot(m, name) -> Pdot(module_path m, name)
+            | Lapply _ -> assert false
+          in
```

Comment:
A bit strange to have to do all this reconstruction by hand.
Could it at least be extracted as independent function?